### PR TITLE
Add the credential environment configuration to the Git checkout command

### PR DESCRIPTION
**Pull request by ivanduplenskikh**
**Description**: The credential environment config property is added to the Git **checkout** command when either filter fetch input or the agent knob `ADD_FORCE_CREDENTIALS_TO_GIT_CHECKOUT` is provided.
This PR addresses the issue of the pipeline agent encountering credentials prompting when attempting to access a private repository with specified filter options.

**Related WI**: [AB#2209290](https://dev.azure.com/mseng/b924d696-3eae-4116-8443-9a18392d8544/_workitems/edit/2209290), #4860

### CODE REVIEW

**Review by DmitriiBobreshev — APPROVED** [review]

**Review by max-zaytsev — APPROVED** [review]

File: src/Agent.Plugins/GitSourceProvider.cs [hunk 1]
```diff
@@ -718,8 +718,10 @@ public async Task GetSourceAsync(
                 await RemoveGitConfig(executionContext, gitCommandManager, targetPath, $"http.proxy", string.Empty);
             }
 
+            List<string> additionalFetchFilterOptions = ParseFetchFilterOptions(executionContext, fetchFilter);
             List<string> additionalFetchArgs = new List<string>();
```
Comment (DergachevE) [inline src/Agent.Plugins/GitSourceProvider.cs:722]: nitpick: you can `var` here and below as variable type is explicitly specified

File: src/Agent.Plugins/GitSourceProvider.cs [hunk 2]
```diff
@@ -1374,6 +1381,55 @@ private async Task<bool> IsRepositoryOriginUrlMatch(AgentTaskPluginExecutionCont
             }
         }
 
+        private List<string> ParseFetchFilterOptions(AgentTaskPluginExecutionContext context, string fetchFilter)
```
Comment (DergachevE) [inline src/Agent.Plugins/GitSourceProvider.cs:1384]: nit: the common guidelines prefer to use less-specialized class (i.e., `IList` or `IEnumerable`)

File: src/Agent.Plugins/GitSourceProvider.cs [hunk 3]
```diff
@@ -1374,6 +1381,55 @@ private async Task<bool> IsRepositoryOriginUrlMatch(AgentTaskPluginExecutionCont
...
+        {
+            if (!AgentKnobs.UseFetchFilterInCheckoutTask.GetValue(context).AsBoolean())
+            {
+                return new();
+            }
+
+            // parse filter and only include valid options
+            List<string> filters = new();
+
+            List<string> splitFilter = fetchFilter.Split('+').Where(filter => !string.IsNullOrWhiteSpace(filter)).ToList();
```
[the remark this answers is not in the document]
Comment (ivanduplenskikh) [reply src/Agent.Plugins/GitSourceProvider.cs]: Thank you. The segment of code was relocated from [here](https://github.com/microsoft/azure-pipelines-agent/pull/4965/files/8257f15dc1d3d5a8a926d7a90dd8ce94d5e1fe6c#diff-15b79029a8d2adb12591d5269c8d928aba1032a6c21548bb4bd5f36f4644ef22L236).
Additionally, I implemented a check to ensure the filter string is neither null nor empty.

File: src/Agent.Plugins/GitSourceProvider.cs (hunk 2 above)
Comment (ivanduplenskikh) [reply src/Agent.Plugins/GitSourceProvider.cs]: Thank you for the advice.
The return type **List** has been changed to **IEnumerable**.

File: src/Agent.Plugins/GitSourceProvider.cs [hunk 4]
```diff
@@ -738,6 +740,11 @@ public async Task GetSourceAsync(
                     string configKey = "http.extraheader";
                     string args = ComposeGitArgs(executionContext, gitCommandManager, configKey, username, password, useBearerAuthType);
                     additionalFetchArgs.Add(args);
+
+                    if (additionalFetchFilterOptions.Count() != 0 && AgentKnobs.AddForceCredentialsToGitCheckout.GetValue(executionContext).AsBoolean())
```
Comment (DergachevE) [inline src/Agent.Plugins/GitSourceProvider.cs:744]: nit: `Any()`?

**Review by DergachevE — APPROVED** [review]

**Pull request merged**